README.md
# MPHP7 Single (Multi PHP - Single Site) Docker Image on mod_php

[![Total Downloads](https://img.shields.io/docker/pulls/idimsh/mphp7s?style=square)](https://hub.docker.com/repository/docker/idimsh/mphp7s)
[![MicroBadger Layers](https://img.shields.io/microbadger/layers/idimsh/mphp7s/latest)](https://hub.docker.com/repository/docker/idimsh/mphp7s)
[![MicroBadger Image Size](https://img.shields.io/microbadger/image-size/idimsh/mphp7s/latest)](https://hub.docker.com/repository/docker/idimsh/mphp7s)


## Why
I am a fan of mod_php with Apache and not much into PHP-FPM, in this image 
Nginx works as reverse proxy to Apache instances which run a specific 
version of PHP using mod_php.

## Purpose
To host a single web-application with the desired PHP version and to be 
able to change the PHP version from the project (web project) document 
root via config file. 
The configuration for Nginx is generated automatically.

## Structure
The image is based on Ubuntu 18.04 and runs Nginx on port 80 which works as 
revers proxy.  
3 Apache instances run on different ports for 3 versions of 
PHP: 7.1, 7.2, 7.3, 7.4.  
Each apache instance operates mod_prefork and listens on 
ports: 83, 84, 85, 86 
respectively, each of which uses a different mod_php version.  
  
A main directory which contains the web application is supposed to 
be mounted to the container r/w to '/srv/' directory which is the 
main DocumentRoot.

## Features:  
In addition to most of the commonly used PHP extensions from Ubuntu,
The following extra PHP Extensions are available:
- xdebug
- inotify
- ev
- apcu
- memcached
- mongodb
- igbinary
- redis
- swoole
  
And they are all compiled (by default) using `pecl` when the image was created.

The compilation of extension is controlled by build argument (`--build-arg`) `BUILD_EXTENSIONS` which is a space separated list of extension names to be installed for every PHP version.    


## Environment variables:
`PHP_VERSION`: (default: 7.4) The version of PHP which serves the website 
and which runs the cli 'php' command.  
Supported values: 7.1, 7.2, 7.3, 7.4 
  
`DOCROOT`: (default: ".") Must me set, it is the relative path inside `'/srv/'`
directory which will be the document root.
  
`LOG_TO_SRV`: (default: "0") If set to non "0", the log files of Apache, Nginx, and PHP will reside in `/srv/system/log/`
directory where each service will have a subdirectory there, otherwise they will be in the default '/var/log/' inside the container.  
Further tests on permissions need to be conducted on Linux docker hosts.  

## Running:
To run and auto start all services:  
`docker run -it idimsh/mphp7s:latest`  
  
It will execute the default supervisor command which starts all services.  
Pressing Ctrl+C in the running container will kill supervisor and 
services running then terminate the container.  
  
Or go to shell by  
`docker run -it idimsh/mphp7s:latest /bin/bash`  
  
Then execute:  
`run.sh`  
To get supervisor running in the foreground. Pressing Ctrl+C will kill 
supervisor but keep all services running and return to shell.  
  
#### Note
Supervisor is not really *controlling* the processes, it just call the start
`service` or `init.d` command, I did not build commands to control Apache or
Nginx using supervisor (to support auto reload, exit status monitoring, 
signals, ...). The implementation of supervisor is simple.  
  
##### Full command:
`docker run --rm -it -v .:/srv --hostname d1.mphp7s.loc -e PHP_VERSION=7.2 -e DOCROOT=public -e LOG_TO_SRV=1 -p 8080:80 idimsh/mphp7s:latest`  
  
`-v .:/srv`: mounts current directory to base directory '/srv/' (which must be '/srv').  
`--hostname d1.mphp7.loc`: Set host name of the container to a FQDN which reduces Apaches' startup warnings.  
`-e PHP_VERSION=7.2`: Set PHP version to 7.2.  
`-e DOCROOT=public`: Set the serving docroot to '/srv/public/', it is a relative directory.  
`-e LOG_TO_SRV=1`: Write log files for services to '/srv/system/log/'.  
  
## Controlling PHP version from PHP Project:  
In the mounted DocRoot to '/srv/', a file named: 'php-version' is read and
the first line of it controls the PHP version, supported values: 7.1, 7.2, 7.3, 7.4.  
  
The existence of this file overrides the PHP version passed in environment 
(`-e` param), moreover: this file is constantly being monitored for changes
using `incron` daemon from the container and any changes to it will cause
the Nginx server to reload and serve the changed PHP version.  
For this to work however, the file '/srv/php-version' should exists when 
the container is starting. (i.e: it is not monitored for creation).  

## Building from Dockerfile:
Build locally by cloning the repository
```bash
git clone git@github.com:idimsh/mphp7s.git .
cd mphp7s
docker build -t "idimsh-mphp7s:latest" .
```  
This will build the image with all extra PHP extensions (the default ones) which are: `xdebug inotify ev apcu memcached mongodb igbinary redis swoole`  
  
To build with custom extensions:
```bash
docker build -t "idimsh-mphp7s:latest" --build-arg BUILD_EXTENSIONS="apcu" .
```  
This will build only with `acpu` extension

## Comsuming the image from docker-compose:  
Here is an example of docker-compose.yaml
```docker
version: '3.5'
services:
  my-web:
    image: idimsh/mphp7s:latest
    container_name: my-web
    hostname: my-web.loc
    args:                                                                      
        BUILD_EXTENSIONS: 'apcu memcached'
    environment:
        PHP_VERSION: 7.3
        DOCROOT: public
    ports:
      - '8082:80'
    volumes:
      - '.:/srv:rw'
    networks:
      default:
        aliases:
          - my-web
          - my-web.loc
```





CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project Do NOT adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.4] - 2020-01-05
### Changed
- Fixed Nginx config and docroot path  
  

## [1.3] - 2019-12-25
### Added
- PHP version 7.4
- Ability to write log files to mounted working directory instead of /var/log/
- Dynamic PHP extension via build args, with a default list of:  
`xdebug inotify ev apcu memcached mongodb igbinary redis swoole` 
- Build Arguments:  
`BUILD_ZEND_EXTENSIONS`  
`BUILD_EXTENSIONS`
- Environment Variables:  
`LOG_TO_SRV`  
- Ability to override PHP ini values via file: `/srv/system/php/php.ini` 
- Ability to override Nginx config values via files:  
`/srv/system/nginx/server.main.conf`: for the main `server` block  
`/srv/system/nginx/location.root.conf`: for `location /` inside main `server`  
`/srv/system/nginx/headers.conf`: for static files `location` inside main `server`  

### Removed
- exim email server


## [1.2] - 2019-05-30
### Added
- Exposed environment variable `PHP_VERSION` in Dockerfile
- Added environment variable `DOCROOT` which is a relative path inside '/srv'
to serve files from

### Changed
Nothing major:
- envvars file for each apache instance is shortened, and now "sourcing" 
one common file: /etc/apache2/envvars-common
- 000-default.conf is generated via template file, the replacement 
(replacing ${DOCROOT} with the value of the env-var defined)
is done in /etc/apache2/envvars-common
- 000-default.conf (default virtual host) is now a single, 
definitions for the 3 apache instances are symlinks to one file, 
previously they were the same copy.
- Supervisor command for apache processes are now using init.d scripts
previously they where calling 'service' wrapper, this is required to pass
environment variables to apache init script.

## [1.1] - 2019-05-27
### Added
- PHP extension: swoole, to all PHP versions

## [1.0] - 2019-05-05
### Added
- Initial Release



scripts/server/set-php-version.sh
#!/bin/bash

SCRIPT_NAME="$(cd $(dirname "$0"); pwd -P)/$(basename "$0")"
SCRIPT_DIR="$(dirname "$SCRIPT_NAME")"
versions_file="${SCRIPT_DIR}/php-versions.txt"
versions=$(head -n1 $versions_file)

version="$1"

for v in $versions; do
  if [ "$version" = "$v" ]; then
    update-alternatives --quiet --set phar /usr/bin/phar${v} && \
    update-alternatives --quiet --set phar.phar /usr/bin/phar.phar${v} && \
    update-alternatives --quiet --set php /usr/bin/php${v} && \
    [ -f /usr/bin/php-config${v} ] && update-alternatives --quiet --set php-config /usr/bin/php-config${v} || true && \
    [ -f /usr/bin/phpize${v} ] && update-alternatives --quiet --set phpize /usr/bin/phpize${v} || true

    echo "updated default PHP version to ${v}"
    exit 0;
  fi
done

echo "unsupported PHP version: '${v}'" >&2
exit 1


scripts/server/install-php-extension.sh
#!/bin/bash
set -e

SCRIPT_NAME="$(cd $(dirname "$0"); pwd -P)/$(basename "$0")"
SCRIPT_DIR="$(dirname "$SCRIPT_NAME")"
VERSIONS_FILE="${SCRIPT_DIR}/php-versions.txt"
VERSIONS=$(head -n1 $VERSIONS_FILE)

extension_name="$1"

if [ -z "$extension_name" ]; then
  echo "extension name must be provided" >&2
  exit 1;
fi

if echo "$extension_name" | grep -q ' '; then
  echo "extension name [$extension_name] must not contain a space, not supported for multiple" >&2
  exit 1;
fi

declare -A VERSIONS_API
declare -A VERSIONS_LIB_PATH
VERSIONS_API["7.1"]="20160303"
VERSIONS_API["7.2"]="20170718"
VERSIONS_API["7.3"]="20180731"
VERSIONS_API["7.4"]="20190902"
VERSIONS_LIB_PATH["7.1"]="/usr/lib/php/20160303"
VERSIONS_LIB_PATH["7.2"]="/usr/lib/php/20170718"
VERSIONS_LIB_PATH["7.3"]="/usr/lib/php/20180731"
VERSIONS_LIB_PATH["7.4"]="/usr/lib/php/20190902"

[ -z "$ZEND_EXTENSIONS" ] && ZEND_EXTENSIONS="xdebug"

for v in $VERSIONS; do
  $SCRIPT_DIR/set-php-version.sh $v
  pecl channel-update pecl.php.net
  echo | pecl install --force $extension_name
  pecl uninstall --register-only $extension_name
  strip ${VERSIONS_LIB_PATH[$v]}/$extension_name.so
  if echo "$ZEND_EXTENSIONS" | egrep -q '(^| )'"$extension_name"'( |$)'; then
    echo "zend_extension=${VERSIONS_LIB_PATH[$v]}/$extension_name.so" > /etc/php/$v/mods-available/$extension_name.ini
  else
    echo "extension=$extension_name.so" > /etc/php/$v/mods-available/$extension_name.ini
  fi
  ln -sf /etc/php/$v/mods-available/$extension_name.ini /etc/php/$v/cli/conf.d/30-$extension_name.ini
  ln -sf /etc/php/$v/mods-available/$extension_name.ini /etc/php/$v/apache2/conf.d/30-$extension_name.ini
done


scripts/server/update-default-website.php
#!/usr/bin/env php
<?php
/**
 * setup an Nginx default server proxy template.
 * Used for docker container created by Abdulrahman Dimashki <idimsh@gmail.com>
 */
define( 'PHP_VERSION', '7.3' );

function print_e( ...$input ) {
  fwrite( STDERR, implode( ' ', $input ) . "\n" );
}

$php_version = getenv( 'PHP_VERSION' );

if ( ! in_array( $php_version, $php_versions = [ '7.1', '7.2', '7.3' ] ) ) {
  $php_version = PHP_VERSION;
  print_e( "PHP version does not match or not given, set to: {$php_version}" );
}

$dest_extra_files = [
  'default_server_proxy' => [
    'template' => '/etc/nginx/default_server.proxy.template',
    'dest'     => '/etc/nginx/default_server.proxy',
  ],
];

foreach ( $dest_extra_files as $files_array ) {
  if ( ! file_put_contents( $files_array['dest'], replace_templates( file_get_contents( $files_array['template'] ) ) ) ) {
    print_e( "error creating file [{$files_array['dest']}]" );
    exit( 1 );
  }
}


/**
 * Variables in default proxy templates:
 * {if-php-7.1}: block
 * {if-php-7.2}: block
 * {if-php-7.3}: block
 * ---
 * User Include Files:
 * /etc/nginx/default_server.proxy.template
 * /etc/nginx/default_server.proxy
 */

/**
 * This is special to replace conditional Blocks depending on condition passed.
 *
 * @param $condition
 * @param $start_string
 * @param $end_string
 * @param $input_string
 *
 * @return null|string|string[]
 */
function preg_replace_conditional( $condition, $start_string, $end_string, $input_string ) {
  $pattern_start = preg_quote( $start_string, '@' );
  $pattern_end   = preg_quote( $end_string, '@' );
  if ( $condition ) {
    $pattern_start = "@{$pattern_start}.*\\n@";
    $pattern_end   = "@{$pattern_end}.*\\n@";
    $input_string  = preg_replace( $pattern_start, '', $input_string );
    $input_string  = preg_replace( $pattern_end, '', $input_string );
  } else {
    // non greedy match which matches new line also with 's' modifier.
    $pattern      = "@{$pattern_start}.*?{$pattern_end}\\n?@s";
    $input_string = preg_replace( $pattern, '', $input_string );
  }

  return $input_string;
}

/**
 * This will replace the content of  the input string moving through all defined
 * template variables
 *
 * @param $input_string
 *
 * @return mixed|null|string|string[]
 */
function replace_templates( $input_string ) {
  global $php_version;

  $output = str_replace( "\r", "", $input_string );

  $output = preg_replace_conditional( $php_version == '7.1', '{if-php-7.1}', '{/if-php-7.1}', $output );
  $output = preg_replace_conditional( $php_version == '7.2', '{if-php-7.2}', '{/if-php-7.2}', $output );
  $output = preg_replace_conditional( $php_version == '7.3', '{if-php-7.3}', '{/if-php-7.3}', $output );

  return $output;
}



scripts/server/replace-templates.php
#!/usr/bin/env php
<?php
/**
 * setup an Nginx default server proxy template.
 * Used for docker container created by Abdulrahman Dimashki <idimsh@gmail.com>
 */
define( 'DEFAULT_PHP_VERSION', '7.4' );

function print_e( ...$input ) {
  fwrite( STDERR, implode( ' ', $input ) . "\n" );
}

$versions = explode( " ", file( __DIR__ . '/php-versions.txt' )[0] );

$PHP_VERSION = getenv( 'PHP_VERSION' );
$LOG_TO_SRV  = getenv( 'LOG_TO_SRV' );
$DOCROOT     = getenv( 'DOCROOT' );

if ( ! in_array( $PHP_VERSION, $versions ) ) {
  $PHP_VERSION = DEFAULT_PHP_VERSION;
  print_e( "PHP version does not match or not given, set to: {$PHP_VERSION}" );
}

$dest_extra_files = [
  'default_server_proxy' => [
    'template' => '/etc/nginx/default_server.proxy.template',
    'dest'     => '/etc/nginx/default_server.proxy',
  ],
  'default_server_conf'  => [
    'template' => '/etc/nginx/default_server.conf.template',
    'dest'     => '/etc/nginx/default_server.conf',
  ],
];

foreach ( $dest_extra_files as $files_array ) {
  if ( ! file_put_contents( $files_array['dest'], replace_templates( file_get_contents( $files_array['template'] ) ) ) ) {
    print_e( "error creating file [{$files_array['dest']}]" );
    exit( 1 );
  }
}


/**
 * Variables in default proxy templates:
 * {if-php-7.1}: block
 * {if-php-7.2}: block
 * {if-php-7.3}: block
 * ---
 * User Include Files:
 * /etc/nginx/default_server.proxy.template
 * /etc/nginx/default_server.proxy
 *
 *
 * Variables in default_server_conf templates:
 * {if-LOG_TO_SRV}: block
 * {if-not-LOG_TO_SRV}: block
 */

/**
 * This is special to replace conditional Blocks depending on condition passed.
 *
 * @param $condition
 * @param $start_string
 * @param $end_string
 * @param $input_string
 *
 * @return null|string|string[]
 */
function preg_replace_conditional( $condition, $start_string, $end_string, $input_string ) {
  $pattern_start = preg_quote( $start_string, '@' );
  $pattern_end   = preg_quote( $end_string, '@' );
  if ( $condition ) {
    $pattern_start = "@{$pattern_start}.*\\n@";
    $pattern_end   = "@{$pattern_end}.*\\n@";
    $input_string  = preg_replace( $pattern_start, '', $input_string );
    $input_string  = preg_replace( $pattern_end, '', $input_string );
  } else {
    // non greedy match which matches new line also with 's' modifier.
    $pattern      = "@{$pattern_start}.*?{$pattern_end}\\n?@s";
    $input_string = preg_replace( $pattern, '', $input_string );
  }

  return $input_string;
}

/**
 * This will replace the content of  the input string moving through all defined
 * template variables
 *
 * @param $input_string
 *
 * @return mixed|null|string|string[]
 */
function replace_templates( $input_string ) {
  global $PHP_VERSION, $LOG_TO_SRV, $DOCROOT;

  $output = str_replace( "\r", "", $input_string );

  $output = preg_replace_conditional( $PHP_VERSION == '7.1', '{if-php-7.1}', '{/if-php-7.1}', $output );
  $output = preg_replace_conditional( $PHP_VERSION == '7.2', '{if-php-7.2}', '{/if-php-7.2}', $output );
  $output = preg_replace_conditional( $PHP_VERSION == '7.3', '{if-php-7.3}', '{/if-php-7.3}', $output );
  $output = preg_replace_conditional( $PHP_VERSION == '7.4', '{if-php-7.4}', '{/if-php-7.4}', $output );

  $output = preg_replace_conditional( ! empty( $LOG_TO_SRV ), '{if-LOG_TO_SRV}', '{/if-LOG_TO_SRV}', $output );
  $output = preg_replace_conditional( empty( $LOG_TO_SRV ), '{if-not-LOG_TO_SRV}', '{/if-not-LOG_TO_SRV}', $output );

  $output = str_replace( '${DOCROOT}', $DOCROOT, $output );

  return $output;
}

